### PR TITLE
added a way to set max enchants to go on an item

### LINKED ISSUE #3447: EcoEnchants & Auto Enchanter -> Exceeding amount of enchantments
### ❗ Checklist

- [X] I am using the official english version of Slimefun and did not modify the jar.
- [x] I am using an up to date "DEV" (not "RC") version of Slimefun.
- [X] I am aware that issues related to Slimefun addons need to be reported on their bug trackers and not here.
- [X] I searched for similar open issues and could not find an existing bug report on this.

### 📍 Description

AlternativeSoap — i dag kl. 06:19
Need some help here...
Does Auto Enchanter got an enchantment limiter? I know it got it inside "items" config, but if I set it to 8 my players can still get over 8 enchantments.

[...]

AlternativeSoap — i dag kl. 06:24
I've got ecoenchants, so they can get like really many enchantments at the same time

BOOMER_1 — i dag kl. 06:25
iirc, you can limit the # of those enchants in their config

variananora — i dag kl. 06:25
it will be bypassed by sf

AlternativeSoap — i dag kl. 06:25
I know, but it doesn't work with the auto enchanter...

variananora — i dag kl. 06:25
using auto enchanter

AlternativeSoap — i dag kl. 06:26
They said I should try to contact you if you got any ideas or workaround

BOOMER_1 — i dag kl. 06:26
ecoenchants shouldn't work in the enchanter should they?  they aren't actual enchants 

variananora — i dag kl. 06:26
hmm
can you show us a vid or smth
i'm a little bit lost
and just open an issue on github

variananora — i dag kl. 06:27
so we can track it properly

BOOMER_1 — i dag kl. 06:27
alternativesoap, are you saying players are able to apply ecoenchantments via the auto enchanter

AlternativeSoap — i dag kl. 06:27
No, I am saying they can apply more enchantments than the hard-cap.

BOOMER_1 — i dag kl. 06:27
of eco enchants or vanilla enchants

AlternativeSoap — i dag kl. 06:27
both

variananora — i dag kl. 06:28
yes, to a degree 

BOOMER_1 — i dag kl. 06:28
he's not saying that

variananora — i dag kl. 06:28
apply to item A
disenchant it
then enchant it on item B

BOOMER_1 — i dag kl. 06:28
he's saying the quantity of both are over limits

AlternativeSoap — i dag kl. 06:28
Or just use "auto enchanter" constant with a enchanted book everytime 

variananora — i dag kl. 06:29
item B exceed the max cap of enchantment configured by eco enchants

BOOMER_1 — i dag kl. 06:29
but ecoenchants should not work in the disenchanter. that's what i'm trying to find out
or the enchanter

AlternativeSoap — i dag kl. 06:29
No, not the disenchanter... Only the auto enchanter

variananora — i dag kl. 06:29
amount of enchantments. not the level of enchantments

BOOMER_1 — i dag kl. 06:29
correct
quantity

AlternativeSoap — i dag kl. 06:30
Yup correct

variananora — i dag kl. 06:30
eco enchanter is one of plugins that register itself as a 'proper' enchantment
it's not like sf 'life steal'
alternativesoap can you open the issue on github?
so we can track it?

Removed some text, but I am too tired and too lazy to explain it on point.

### 📑 Reproduction Steps

Use ecoenchants, use hard-cap quantity of enchantments on one tool at a time example 4, use an auto enchanter, go over the limit with that.

### 💡 Expected Behavior

Not to exceed the limit, everything else work but not the auto enchanter...

### 📷 Screenshots / Videos

_No response_

### 📜 Server Log

_No response_

### 📂 `/error-reports/` folder

_No response_

### 💻 Server Software

Paper

### 🎮 Minecraft Version

1.18.x

### 🧭 Other plugins

EcoEnchants

### PR DESCRIPTION
**Pull request by J3fftw1**
## Description

This pr adds a way to sets a limit of how many enchants can go on an item.
Default is false
Default of the max enchants when its enabled is 10.

## Proposed changes

Added an item setting that checks for a limit on how many enchants can go on an item.

## Related Issues (if applicable)

Resolves #3447 

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [x] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I have made sure that the proposed changes do not break compatibility across the supported Minecraft versions (1.14.* - 1.19.*).
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [ ] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/enchanting/AbstractEnchantmentMachine.java [hunk 1]
```diff
@@ -84,4 +90,8 @@ protected boolean hasIgnoredLore(@Nonnull ItemStack item) {
 
         return false;
     }
+
+    protected boolean IsEnchantmentAmountAllowed(@Nonnull ItemStack item ) {
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/enchanting/AbstractEnchantmentMachine.java:94]:
```suggestion
    protected boolean isEnchantmentAmountAllowed(@Nonnull ItemStack item ) {
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/enchanting/AutoEnchanter.java [hunk 2]
```diff
@@ -109,13 +109,21 @@ protected MachineRecipe enchant(BlockMenu menu, ItemStack target, ItemStack ench
         }
 
         /*
-         * If override is false, remove those with lower level so we don't override existing enchants
-         * This also removes those with the same level so they aren't accounted for enchanting time
+         * If override is false, remove those with lower level, so we don't override existing enchants
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/enchanting/AutoEnchanter.java:112]: This and line below were fine before

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/enchanting/AutoEnchanter.java [hunk 3]
```diff
@@ -109,13 +109,21 @@ protected MachineRecipe enchant(BlockMenu menu, ItemStack target, ItemStack ench
...
          */
         if (!overrideExistingEnchantsLvl.getValue()) {
             enchantments.entrySet().removeIf(e -> target.getEnchantmentLevel(e.getKey()) >= e.getValue());
         }
 
+        /*
+         * When maxEnchants is set to -1 it will be ignored. When it's set to 0 it will not allow any enchants to go
+         * on an item. When maxEnchants is set to any other value it will allow that many enchants to go on the item.
+         */
+        if (IsEnchantmentAmountAllowed(target)) {
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/enchanting/AutoEnchanter.java:123]:
```suggestion
        if (isEnchantmentAmountAllowed(target)) {
```

**Review by WalshyDev — CHANGES_REQUESTED** [review]
Disenchanter?

**Comment by J3fftw1** [comment]
> Disenchanter?

This is for how many enchants can go on an item.
Doesnt apply for Disenchanter imo.
Can add it if thats preferred tho

**Comment by Boomer-1** [comment]
doesn't really make sense to add it to the disenchanter.

**Comment by WalshyDev** [comment]
Don't see why not. It removes enchants. We do the same for enchantment level and generally try to keep them in sync when it comes to features/settings.

**Comment by Boomer-1** [comment]
On the disenchanter, then it is set one of two ways. You can only remove an enchantment if you have less than a preset number, or if you have above a certain number. While I'd agree normally you want both the same, this one just doesn't make sense on the disenchanter. You could add it with the default of it not being applied and server owners can choose. I just don't see a use case scenario for it.

**Comment by J3fftw1** [comment]
👀

**Comment by JustAHuman-xD** [comment]
Is there any display that shows this limit to the player? Otherwise I feel like that will make a lot of player confused.

**Comment by J3fftw1** [comment]
> Is there any display that shows this limit to the player? Otherwise I feel like that will make a lot of player confused.

no but that isnt the case with any limit we apply in slimefun

**Review by Sfiguz7 — APPROVED** [review]
LGTM

**Review by svr333 — APPROVED** [review]

**Pull request merged**